dissect/xfs/xfs.py

```diff
@@ -450,8 +450,8 @@ def datafork(self):
         else:
             size = self.ag.sb.sb_inodesize - offset
 
-        if self.inode.di_forkoff:
-            size -= size - self.inode.di_forkoff * 8
+            if self.inode.di_forkoff:
+                size -= size - self.inode.di_forkoff * 8
```

Comment:
This shouldn't be indented, `XFS_DINODE_FMT_LOCAL` inodes can still have an attribute fork.

And isn't this the same as:
```suggestion
        if self.inode.di_forkoff:
            size = self.inode.di_forkoff * 8
```

Reply:
That doesn't matter, `XFS_DINODE_FMT_LOCAL` inodes have a fixed size at `self.size`, whereas for other types you need to calculate the size yourself. The datafork size is the area between the inode structure and the start of the attribute fork.

Also I don't think that's the same. I haven't checked but it should probably be closer to:
```python
size = (self.inode.di_forkoff * 8) - self.ag.sb.sb_inodesize - offset
```

Reply:
Ah, good, I didn't catch that.

The size should be as I mentioned, according to the specs:

> The location of the attribute fork in the inode's literal area (offset 100 to the end of the inode) is
> specified by the di_forkoff value in the inode's core. If this value is zero, the inode does not
> contain any extended attributes. Non-zero, the byte offset into the literal area = di_forkoff * 8

And the literal area starts with the datafork, so no need to calculate wrt the offset from the start of the  inode structure.

Reply:
First time for me to dive into XFS... but should it not be more like:

```
if self.inode.di_format == c_xfs.xfs_dinode_fmt.XFS_DINODE_FMT_LOCAL:
    size = self.size
elif self.inode.di_forkoff:
    size = self.ag.sb.sb_inodesize - (self.inode.di_forkoff * 8) - offset
else:
    size = self.ag.sb.sb_inodesize - offset
```

Reply:
and.. maybe we should add that part of the docs as well...?

Reply:
Did a quick check, @pyrco is indeed correct. So the logic should look like this:

```python
    def datafork(self):
        offset = 0xB0 if self.inode.di_version == 0x3 else 0x64
        if self.inode.di_format == c_xfs.xfs_dinode_fmt.XFS_DINODE_FMT_LOCAL:
            size = self.size
        elif self.inode.di_forkoff:
            size = self.inode.di_forkoff * 8
        else:
            size = self.ag.sb.sb_inodesize - offset

        return RangeStream(self._buf, offset, size)
```

Looking into it, I also now believe the `attrfork` function to be faulty. However since it's not actually currently used anywhere (we don't parse xattr yet) it wasn't apparent. I think it should be this:
```python
    def attrfork(self):
        if self.inode.di_forkoff == 0:
            raise Error(f"{self!r} has no extended attributes")

        offset = 0xB0 if self.inode.di_version == 0x3 else 0x64
        offset += self.inode.di_forkoff * 8
        size = self.ag.sb.sb_inodesize - offset

        return RangeStream(self._buf, offset, size)
```